Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityRaycasterProvider.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Physics;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// The resulting hit information from an IMixedRealityRaycasterProvider.
+    /// </summary>
+    public struct RaycasterHit
+    {
+        public float Distance;
```

Comment:
I'm confused as to why we need a RaycastHit-analogue since it doesn't add anything new. Why not stick with RaycastHits?

This struct omits a bunch of useful fields like the collider, the rigidbody. triangle index, texture coords - I use this stuff all the time.

Reply:
Good point, those should be added back to have parity with the existing RaycastHitInfo from Unity. Will do that in a future iteration. 

I'm also consider using the Unity property names from RaycastHitInfo exactly, like 'distance' instead of 'Distance', to make the change less breaking.

Reply:
And more broadly: The overall reason for the new RaycasterHit struct is that Unity's RaycastHitInfo is not extensible. Only the position, normal, and distance can be set on it, and important properties that MRTK relies on, like the 'transform' or 'collider', are not settable.

This struct allows for those types to be specified.

Reply:
>This struct allows for those types to be specified.

Ah that makes sense.

In that case, maybe we should go with an extendable struct with a 1:1 parity, including assignment overrides so you can assign all the values from a RaycastHit to a RaycasterHit with one = operation and/or return a RaycastHit as a RaycasterHit.

Reply:
Good that this change is happening. We would also want to extend the struct to add additional hit data apart from the unity hit data. This change would be really helpful.

Reply:
To make the hit data extensible, the simplest thing to do would be making the RaycasterHit struct a class so that it can be extended and eventually casted back to the type you expect. The downside in doing that is now the allocation is managed by the GC.

The other option I see is to add some sort of property bag to the struct, which I would do if there was some previous example of this in MRTK. I haven't seen that concept used yet though.

I'm leaning towards simply transforming this struct to a class, so that you can extend and cast it to the expected type. Let me know if that works or if there are other suggestions.

Reply:
Is there a way that we could provide our own struct similar to raycaster? Changing it to class is going to be bad on GC as the objects are created in Update loop. If the struct is changed to class, then caching and reusing the references need to be done.

Reply:
While not ideal from the garbage collector perspective, I don't have a better idea to solve this. I'm not going to implement pooling/caching for this type either because it's unlikely the potential additional GC pressure will be an issue, especially given that the event data generated by MRTK is all class based as well.

I'm open to other approaches to make this more performant, but using a class here seems like the best trade-off in order to extend this with more data.

That, or we disallow extension of this type and leave it as a struct. It sounds like for your scenario you plan on implementing a custom FocusProvider as well? Maybe there is some way you can get the additional data from the raycast without having to extend this class. Could it be queried from the GameObject/component directly?

Reply:
Though currently I have written a custom FocusProvider, it would be good if we use raycaster class. But right now, I added a property in FocusProvider for additional hit data. The data that we use is more 3D model specific like triangleId, instanceId etc. We cannot use gameobjects, because we render the game objects by doing mesh combining, whereas our hit happens to individual meshes in the model. 
For now, class could be used. I would update if I could think of a better approach. Thanks!